keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,177 @@
+---
+title: Immutable Fields
+authors:
+  - "@apelisse"
+owning-sig: sig-api-machinery
+participating-sigs:
+  - sig-api-machinery
+reviewers:
+approvers:
+editor: TBD
+creation-date: 2019-06-03
+last-updated: 2019-06-03
+status: provisional
+see-also:
+  - "/keps/sig-api-machinery/0006-apply.md"
+---
+
+# Immutable fields
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+A lot of fields in APIs tend to be "immutable", they can't be changed after
+creation. This is true for example for many of the fields in pods. There is
+currently no way to declaratively document that fields are immutable, and one
+has to rely on either built-in validation for core types, or have to build a
+validating webhooks for CRDs.
+
+Providing a new `// +immutable` marker would make the API more descriptive to
+users and help API developers by verifying these assertions automatically.
+
+## Motivation
+
+The kubernetes model tend to promote "immutable" objects rather than "delete and
+recreate" on changes approach, and users have expressed the need for such a
+feature through this issue for example:
```

Comment:
I find the first sentence very confusing, what is the conveyed message? That API objects fields can currently be modified without having to delete and re-create the object?

Reply:
I'll rephrase :-)

Reply:
tentatively rephrased